src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,264 @@
+//  - Pro: Random selection or fixed budget would do the trick of adaptively upgrading the tree compression over time,
+//  - Cons: Requires to have a task to perform the dictionary creation outside of the insertion process,
+//  - Cons: Compression ratio will not be THAT tight per single page,
+//  - Cons: We may need larger dictionaries to accomodate better compression ratios. 
+//
+//  In Voron we use global scopes dictionaries because they have a cleaner implementation with less intrusive processes
+//  where it really matters (writing). 
+
+unsafe partial class CompactTree
+{
+    public struct FullDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly CompactTree _tree;
+        private Iterator _iterator;
+
+        public FullDictionaryKeyScanner(CompactTree tree)
+        {
+            _tree = tree;
+            _iterator = _tree.Iterate();
+            _iterator.Reset();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.Iterate();
+            _iterator.Reset();
+        }
+    }
+
+    public struct RandomDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly CompactTree _tree;
+        private readonly int _samples;
+        private readonly int _seed;
+        private RandomIterator _iterator;
+
+        public RandomDictionaryKeyScanner(CompactTree tree, int samples, int seed = 0)
+        {
+            _tree = tree;
+            _seed = seed;
+            _samples = samples;
+            _iterator = _tree.RandomIterate(samples, seed: seed);
+            _iterator.Reset();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.RandomIterate(_samples, seed: _seed);
+            _iterator.Reset();
+        }
+    }
+
+    public struct RandomBranchDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly long _samples;        
+        private readonly Random _generator;
+        private readonly CompactTree _tree;
+        private int _currentSample;
+        private int _currentIdx;
+        private IteratorCursorState _cursor;
+        private HashSet<long> _pagesSeen;
+
+        public RandomBranchDictionaryKeyScanner(CompactTree tree, long samples, int seed = 0)
+        {
+            _tree = tree;
+            _samples = samples;
+            _currentSample = 0;
+            _currentIdx = 0;
+            _generator = (seed == 0) ? new() : new(seed);
```

Comment:
Good point on the `0` number.
What kind of error do you think you can get there?

Reply:
I see that we use the random to get a position of an entry from the page of `CursorState`. So maybe some errors related to the actual value of that entry?

Reply:
`0` as a seed is actually an invalid seed in practice. The reason is because when the seed is zero there is a possibility that the entropy for those random number generators that can handle it (some can't) is diminished. 

Digression aside, we dont really care about entropy here, we just need a pseudo-random path so it is mostly a matter of taste. I don't know if we would be able to do anything even knowing the actual seed, because unless you haven't written anything, the next time (most likely) you are not going to go down the same path. From the point of view of debugging you will be doing something like this anyways (which is what I do to figure out this stochastic error types).

```
for i to range(N)
    run_test(seed: i)
```

Whatever we decide to do is fine. If you like `null`, we do null I have absolutely no preference for this (as the technical reason is not applicable).